nixos/modules/services/networking/firewalld/lib.nix

```diff
@@ -0,0 +1,55 @@
+{ lib }:
+
+let
+  inherit (lib) mkOption;
+  inherit (lib.types)
+    either
+    enum
+    nullOr
+    port
+    submodule
+    ;
+  mkPortOption =
+    {
+      optional ? false,
+    }:
+    mkOption {
+      type =
+        let
+          type = either port (submodule {
+            options = {
+              from = mkOption { type = port; };
+              to = mkOption { type = port; };
+            };
+          });
+        in
+        if optional then (nullOr type) else type;
+      description = "";
+      apply =
+        value: if builtins.isAttrs value then "${toString value.from}-${toString value.to}" else value;
```

Comment:
Doing this normalisation in the option's `apply` is convenient.

However, I wonder whether users would be confused if they _read_ the final value, expecting to find a list of ints and attrs, but instead find a list of ints and strings.

E.g. they might read this value to re-use it in another option, or for a `mkIf` condition, etc.

It may be less confusing if we do this stringification later when consuming the option.

Reply:
I wonder how useful would this be.  The typical use would be to use the values from the configurations of some services here, which means that if a user wants to use the ports, then it's better to consult the services' settings directly.